### ISSUE TITLE
Date jumps back to current month

### ISSUE THREAD

**Issue by dtbell99** [wojtekmaj/react-date-picker#48]
When i go back to last month, or last year, choose a date, the calendar jumps back to current month.

**Comment by dtbell99** [comment]
I recorded a video of this occurring if someone wants me to send it to them.

**Comment by dtbell99** [comment]
[react-date-picker-bug.zip](https://github.com/wojtekmaj/react-date-picker/files/1692165/react-date-picker-bug.zip)

**Comment by wojtekmaj** [comment]
Hey @dtbell99,
Not sure what's entirely going on here as I don't see your part of the code, but I assume you must have done something wrong with passing the new value to react-date-picker. I assume so because:

* react-date-picker clearly updated itself
* react-date-picker, when updated by changing props, ensures the view it displays includes the value that is chosen
* the date chosen on the calendar remains the same despite you clicking another date

Please mind that react-date-picker does not keep its value, it relies on you handling this in your own state. So it's just like any other input component, actually: you pass `value` prop, you must update it using `onChange`.

**Comment by dtbell99** [comment]
It looks like the picker updates visually but when I console.log in my update message I don't see anything. When I choose a date in the current month I do see console logging. So the update process doesn't hit my onChange code. The value is being set with this.props.object.value so I will see if an inner object is to blame but that is a guess. The other question might be date format perhaps and timezone.. 

Here is the UI declaration.
<DatePicker name="startDate" onChange={(value) => this.props.updateDate(value, "startDate")} value={this.props.service.startDate}/>

Here is the update method
updateDate = (date, type) => {
    console.log('DATE IS BEING UPDATED:'+date+' type:'+type)
    var service = this.state.service
    if (type === "startDate") {
      service.startDate = date
    } else if (type === "endDate") {
      service.endDate = date
    }
    this.setState({
      service
    })
  }

Service is just a object with a startDate date field. 

"react-date-picker": "^6.7.0",
"moment": "^2.20.1",
"react": "^16.2.0",

I'm going to record another video showing the update call being made and not made hopefully that helps. 

I'm also going to create a new Greenfield app with these settings and see if it works, I'll send you the GitHub repo if I can recreate it in that new project.

**Comment by dtbell99** [comment]
[react_date_picker_second_attempt_480.zip](https://github.com/wojtekmaj/react-date-picker/files/1692815/react_date_picker_second_attempt_480.zip)

**Comment by dtbell99** [comment]
Attached video shows OnChange firing only for current month.

**Comment by wojtekmaj** [comment]
Hmmm... From what I see, your method `updateDate` is wrong. I don't know if that's the only issue we're having here, but you should fix it either way.

You should not mutate objects from state on your own. You did something like this:

* Took an object from state
* Mutated a property in it
* Ordered React to setState and update `service` object

The thing is, you actually did not do the last step. `service` object is still `service` object, so it is not guaranteed React is going to do something about it, as from its perspective, it's the same value. Your update method could look like so:

```js
updateDate = (date, type) => {
  console.log('DATE IS BEING UPDATED:'+date+' type:'+type);
  var nextService = Object.assign({}, this.state.service);
  if (type === "startDate") {
    nextService.startDate = date;
  } else if (type === "endDate") {
    nextService.endDate = date;
  }
  this.setState({
    service: nextService
  });
}
```

or better yet, because objects in state is generally a bad idea:

```js
updateDate = (date, type) => {
  console.log('DATE IS BEING UPDATED:'+date+' type:'+type);
  this.setState({ [type]: date });
}
```

but that would require you to do some work around this function so that your service object is gone and you save the values directly in state.

What would also help with pinpointing the issue would be, right next to date picker, putting a date as a text - eg. `this.props.service.startDate.toLocaleDateString()`. That would give us an idea if the dates are even passed to DatePicker or not.

**Comment by dtbell99** [comment]
Good evening,

I made all the changes you suggested and I still get the exact same result. Doing further testing against the control I am seeing that I can't switch years either. 

I think what is best to verify is for me to create a brand new project and only create one component with the date picker and see if it works ok with all the other libraries in my package.json. Then I'll start adding in pieces one by one until I can break it again similar to how we handled PC hardware problems back in the day. It is very possible the problem is the date object which is coming from postgresql using pg in node. Since JSON doesn't recognize dates I use following code to turn it into a javascript date. When I console out the data it is correct as far as my eyes can tell. 


const dateFormat = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z$/;
function stringToDate(value) {
    if (typeof value === "string" && dateFormat.test(value)) {
        return new Date(value);
    }
    return value;
}


Very strange though that i can't move beyond the month/year even if that isn't what is in the field. 

If you have any other ideas for me to try I'll give it a go otherwise tomorrow during the day I'll greenfield an app and see what I can figure out. 

Thanks,
David

**Comment by wojtekmaj** [comment]
Hey @dtbell99, I'll do my best to help you. Thanks for the intention to create greenfield app - it's making it much easier. Although please mind issue #49 :)

**Comment by dtbell99** [comment]
Hi; I haven't gotten time to create a Greenfield app yet. I did do some cross browser testing and the data picker does work in Safari 11.0.3 on macOS High Sierra 10.13. It doesn't not work in Chrome on Windows or macOS. So the fact it does work on Safari is very strange, but again I'll see if I can get it working on Chrome with a clean build which I'll make public if I can make it not work.

**Comment by dtbell99** [comment]
Ok, Greenfield app works in all browsers. My app when going back a month or year the event isn't registered as a change so no feedback. In Greenfield all dates do register and call handler. My next thought is a conflicting library. I'm going to see if there is maybe a conflict with another module.

**Comment by dtbell99** [comment]
I have created a new app from scratch with same modules without the ability to reproduce. My guess would be possible CSS issues or some other loaded module / code. One thing I noticed from the original app and the greenfield was the version number. My app was using 6.7.0 but the newest app was using 6.8.1. When I deleted node_modules and installed version 6.8.1 my application is suddenly working as expected. Therefore I have to assume whatever bug is firing has been corrected and therefore I'm closing out the ticket. If you want more info let me know. 

Thanks,
David
